Add ignore-cert flag option

bck01215 opened the pull request:
Adds a flag with the option to ignore the certificate

bck01215 commented:
This should resolve issue #2

wbh1 commented on collector/collector.go:
```diff
@@ -73,26 +75,27 @@ type AttunityCollector struct {
 
 	// This is required to force HTTP/1.1
-	// Attunity for some reason advertises HTTP/2 despite not supporting it.
+	// Attunity for some reason advertises HTTP/2 despite not supporting it. :(
```
👍🏻

wbh1 commented on collector/collector.go:
```diff
@@ -73,26 +75,27 @@ type AttunityCollector struct {
 // It should be registered to a Prometheus Registry.
-func NewAttunityCollector(cfg *Config) *AttunityCollector {
+func NewAttunityCollector(cfg *Config, ignoreCert bool) *AttunityCollector {
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: ignoreCert}
```
I don't think that this line would actually work? I'm a little confused as to why you have to type cast it.

wbh1 replied:
Maybe consider setting the `Transport` field of the `client` in here with the TLS option: https://github.com/wbh1/attunity_exporter/blob/6b2367ee774cd372f9d5206035acf61fc7c1c8c1/collector/collector.go#L80-L84

Then change this bit to just modify `client.Transport.TLSNextProto` instead of creating a new `http.Transport` and assigning it. https://github.com/wbh1/attunity_exporter/blob/6b2367ee774cd372f9d5206035acf61fc7c1c8c1/collector/collector.go#L86-L94

wbh1 commented on attunity_exporter.go:
```diff
@@ -11,15 +11,16 @@ import (
+	port       = kingpin.Flag("port", "the port to bind the exporter to.").Default("9103").String()
+	cfgFile    = kingpin.Flag("config-file", "path the config.yml file").Default("config.yml").String()
+	debug      = kingpin.Flag("debug", "Enable debug logging").Bool()
+	ignoreCert = kingpin.Flag("ignore-cert", "Will ignore certificate warnings").Bool()
```
I'd prefer this was a directive in the config file, but also I don't think I'll ever have to use this project again, so have at it 😁

wbh1 replied:
This flag can be dropped now.

wbh1 commented:
Not sure why the Included Tags bit wasn't in this repo before. But it looks like the logic to actually use it isn't there? Something to be aware of.

bck01215 commented:
Updated included tags filter. ALso just a general re-write that I believe you had already done locally. I probably should have made that a separate PR... I also moved ignore_cert into the config field per your advice. I updated the example

wbh1 commented on testdata/config.good.yml:
```diff
@@ -2,6 +2,7 @@ user: "***REMOVED***"
 password: "replacethis"
 weblink: "***REMOVED***"
 timeout: 5
+ignore_cert: yes
```
`s/yes/true/g`

wbh1 commented on config.yml.example:
```diff
@@ -2,6 +2,7 @@ user: ""
 password: ""
 weblink: ""
 timeout: 10
+ignore_cert: yes
```
`s/yes/true/g`

wbh1 commented on collector/tasks.go:
```diff
@@ -231,13 +249,13 @@ func filterExcluded(tl *[]task, excluded []string) []task {
 		}
 	}
-	return filtered
+	*tl = filtered
```
Smart 🧠

wbh1 commented on collector/tasks.go:
```diff
@@ -257,5 +275,29 @@ func filterIncluded(tl *[]task, included []string) []task {
+
+}
+
+func filterIncludedTags(tl *[]task, tags []*string) {
```
This looks like it works 🤷🏻‍♂️

wbh1 commented on collector/collector.go:
```diff
@@ -78,29 +84,56 @@ type AttunityCollector struct {
+
+}
+
+// This is required to force HTTP/1.1
```
This comment is duplicated below. Should probably be removed here.

wbh1 commented on collector/collector.go:
```diff
@@ -64,40 +67,73 @@ type AttunityCollector struct {
 // It should be registered to a Prometheus Registry.
-func NewAttunityCollector(cfg *Config) *AttunityCollector {
+func NewAttunityCollector(cfg *Config, ignoreCert bool) *AttunityCollector {
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: ignoreCert}
```
This line still needs removed and the additional `ignoreCert` arg to the function definition isn't needed anymore.

wbh1 commented on collector/collector.go:
```diff
@@ -64,40 +67,73 @@ type AttunityCollector struct {
-		}
-		client.Transport = transport
-	}
+		transport.TLSNextProto = make(map[string]func(authority string, c *tls.Conn) http.RoundTripper)
```
LGTM

wbh1 commented on collector/collector.go:
```diff
@@ -143,7 +171,7 @@ func validateConfig(conf Config) (err []error) {
 // Describe implements the Describe method of the Prometheus Collector interface
 func (a *AttunityCollector) Describe(ch chan<- *prometheus.Desc) {
-
+	// Hi I do nothing
```
:feelsbadman:

wbh1 commented on attunity_exporter.go:
```diff
@@ -39,7 +41,7 @@ func main() {
 	}
 
-	prometheus.MustRegister(collector.NewAttunityCollector(&cfg))
+	prometheus.MustRegister(collector.NewAttunityCollector(&cfg, *ignoreCert))
```
This arg can be dropped now.

wbh1 commented:
@bck01215 OK - I think this should be the last round 😛 . I obviously don't have a way to test this anymore, so I'm just presuming the logic changes work lol

bck01215 commented:
Cleaned up what I had blindly forgotten. I still prefer using "yes" to "true," but I will listen to the great oracle. 

This has been tested. Not only does it work like the last push, but it works without including unnecessary stuff.

wbh1 commented:
> Cleaned up what I had blindly forgotten. I still prefer using "yes" to "true," but I will listen to the great oracle.
> 
> This has been tested. Not only does it work like the last push, but it works without including unnecessary stuff.

Didn't realize that `yes` was supported, honestly. (It is, among other options)

https://github.com/go-yaml/yaml/blob/7649d4548cb53a614db133b2a8ac1f31859dda8c/resolve.go#L37-L42
